go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return &tempLocalObject{
+		path: p,
+		f:    f,
+	}, nil
+}
+
+func downloadParts(ctx context.Context, objectName, namespace, etag string, requestedSize, totalSize, maxBatchSize, offSet, minPartSize, maxPartNum int64, concurrentListeners int, downloadF downloadFunc) ([]*toDownload, error) {
+	numParts, partSize := getNumPartsAndPartSize(requestedSize, minPartSize, maxPartNum)
+	completedPartChan := make(chan *toDownload, numParts)
+	completedParts := make([]*toDownload, numParts)
+
+	batch := make([]*toDownload, 0)
+	batchSize := int64(0)
+	partNum := 1
+	currentOffSet := offSet
+
+	for partNum <= numParts {
+		if batchSize >= maxBatchSize {
+			err := downloadBatch(ctx, objectName, namespace, etag, batch, completedPartChan, concurrentListeners, downloadF)
+			if err != nil {
+				close(completedPartChan)
+				return nil, err
+			}
+
+			batchSize = 0
+			batch = make([]*toDownload, 0)
+			continue
+		}
+
+		batch = append(batch, &toDownload{
+			partNum: partNum,
+			br:      NewBlobRange(currentOffSet, partSize),
+		})
+
+		batchSize += partSize
+		partNum++
+		currentOffSet = int64(math.Min(float64(partSize)+float64(currentOffSet), float64(totalSize)-float64(1)))
+	}
+
+	if batchSize > 0 && len(batch) > 0 {
+		err := downloadBatch(ctx, objectName, namespace, etag, batch, completedPartChan, concurrentListeners, downloadF)
+		if err != nil {
+			close(completedPartChan)
+			return nil, err
+		}
+	}
+
+	close(completedPartChan)
+
+	idx := 0
+	for cp := range completedPartChan {
+		completedParts[idx] = cp
+		idx++
+	}
+
+	sort.Slice(completedParts, func(i, j int) bool {
+		return completedParts[i].partNum < completedParts[j].partNum
+	})
+
+	return completedParts, nil
+}
+
+func downloadBatch(ctx context.Context, objectName, namespace, etag string, batch []*toDownload, completedPartsChan chan *toDownload, concurrentListeners int, downloadF downloadFunc) error {
+	batchChan := make(chan *toDownload, len(batch))
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	for i := 0; i < concurrentListeners; i++ {
+		eg.Go(func() error {
+			for {
+				select {
+				case <-egCtx.Done():
+					return nil
+				case d, ok := <-batchChan:
+					if !ok {
+						return nil
+					}
+
+					cp, err := downloadF(egCtx, objectName, namespace, etag, d.partNum, d.br)
+					if err != nil {
+						return err
+					}
+
+					completedPartsChan <- cp
+				}
+			}
+		})
+	}
+
+	eg.Go(func() error {
+		defer close(batchChan)
+
+		for {
+			select {
+			case <-egCtx.Done():
+				return nil
+			default:
+				for _, d := range batch {
+					batchChan <- d
+				}
+				return nil
+			}
+		}
+	})
+
+	return eg.Wait()
+}
+
+func uploadParts(ctx context.Context, objectName, uploadID string, numParts, concurrentListeners int, totalSize, maxBatchSize int64, reader io.Reader, uploadF uploadFunc) ([]objectstorage.CommitMultipartUploadPartDetails, error) {
+	completedPartChan := make(chan objectstorage.CommitMultipartUploadPartDetails, numParts)
```

Comment:
Again, I don't know if need much in the way of channels here. Just make the plan, and then iterate of the batches. For each batch, just make the byte slice, read it fully into memory, and then call `eg.Go()` to upload it. Use `eg.SetLimit()` to put a limit on the number of concurrent uploads and to keep memory from blowing out.

Reply:
im not really sure what you mean here either... how do i do this without channels in a way that lets me collect all the etags of the uploaded parts in order to call CommitMultipartUpload?